File: cortex-standalone/train/tokenizer.py
"""
train/tokenizer.py

The Python (training-time) copy of the tokenizer contract (docs/CONTRACTS.md §3). Three
copies of this algorithm exist on purpose — the TypeScript core, this trainer, and the
browser widget — because each runtime can't import the others. All three are pinned
against one shared fixture (`tests/fixtures/tokenizer.parity.json`), and `test_tokenizer.py`
here is the Python half of that pin.

The algorithm, byte-for-byte:

  1. Lowercase (if `lower`) — Python `str.lower()` / JS `toLowerCase()`.
  2. Split into maximal runs of Unicode Letter or Number codepoints. Python checks
     `unicodedata.category(ch)[0] in ("L", "N")`; JS checks `/[\\p{L}\\p{N}]/u`. Both
     target the same Unicode General Category groups. Everything else is a separator and
     is discarded.
  3. Map each word to `vocab[word]`, else `unkId`.
  4. Truncate to `maxLen`, pad with `padId`.

Standard library only — no `regex` package, no NLP dependency — so parity with the other
two copies is a matter of the same small algorithm, not a shared library version.
"""

from __future__ import annotations

import json
import unicodedata
from collections import Counter
from dataclasses import dataclass
from typing import Dict, Iterable, List, Sequence

PAD_TOKEN = "<pad>"
UNK_TOKEN = "<unk>"
PAD_ID = 0
UNK_ID = 1
TOKENIZER_VERSION = "1"


def is_word_char(ch: str) -> bool:
    """True for any Unicode Letter (category `L*`) or Number (category `N*`) codepoint —
    the Python spelling of `/[\\p{L}\\p{N}]/u` (CONTRACTS.md §3.1)."""
    return unicodedata.category(ch)[0] in ("L", "N")


def word_split(text: str) -> List[str]:
    """Split `text` into maximal runs of word characters, discarding everything else."""
    tokens: List[str] = []
    cur: List[str] = []
    for ch in text:
        if is_word_char(ch):
            cur.append(ch)
        elif cur:
            tokens.append("".join(cur))
            cur = []
    if cur:
        tokens.append("".join(cur))
    return tokens


@dataclass
class TokenizerConfig:
    version: str
    lower: bool
    max_len: int
    pad_id: int
    unk_id: int
    vocab: Dict[str, int]
    pad_token: str = PAD_TOKEN
    unk_token: str = UNK_TOKEN

    def to_json_dict(self) -> dict:
        """The exact `tokenizer.json` shape from CONTRACTS.md §3.3."""
        return {
            "version": self.version,
            "lower": self.lower,
            "maxLen": self.max_len,
            "padId": self.pad_id,
            "unkId": self.unk_id,
            "padToken": self.pad_token,
            "unkToken": self.unk_token,
            "vocab": self.vocab,
        }

    @staticmethod
    def from_json_dict(d: dict) -> "TokenizerConfig":
        return TokenizerConfig(
            version=str(d["version"]),
            lower=bool(d["lower"]),
            max_len=int(d["maxLen"]),
            pad_id=int(d["padId"]),
            unk_id=int(d["unkId"]),
            vocab=dict(d["vocab"]),
            pad_token=d.get("padToken", PAD_TOKEN),
            unk_token=d.get("unkToken", UNK_TOKEN),
        )


def build_vocab(utterances: Iterable[str], lower: bool = True) -> Dict[str, int]:
    """Deterministically build a word -> id vocab from the TRAIN split. Ids start at 2
    (0/1 are reserved for `<pad>`/`<unk>`) and are assigned by (descending frequency, then
    alphabetical) — sorted explicitly, never relying on dict/Counter iteration order
    (CONTRACTS.md §3.3)."""
    counts: Counter[str] = Counter()
    for u in utterances:
        text = u.lower() if lower else u
        counts.update(word_split(text))

    ordered = sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))
    vocab: Dict[str, int] = {}
    next_id = 2
    for word, _count in ordered:
        vocab[word] = next_id
        next_id += 1
    return vocab


def build_tokenizer_config(utterances: Sequence[str], max_len: int, lower: bool = True) -> TokenizerConfig:
    vocab = build_vocab(utterances, lower=lower)
    return TokenizerConfig(
        version=TOKENIZER_VERSION,
        lower=lower,
        max_len=max_len,
        pad_id=PAD_ID,
        unk_id=UNK_ID,
        vocab=vocab,
    )


def tokenize(text: str, config: TokenizerConfig) -> List[int]:
    """text -> a fixed-length `config.max_len` list of ids (padded with `config.pad_id`,
    truncated at `config.max_len`, OOV words mapped to `config.unk_id`). MUST match the
    core's `tokenize()` exactly for the same `(text, config)`."""
    lowered = text.lower() if config.lower else text
    words = word_split(lowered)
    ids = [config.vocab.get(w, config.unk_id) for w in words[: config.max_len]]
    while len(ids) < config.max_len:
        ids.append(config.pad_id)
    return ids


def load_tokenizer_config(path: str) -> TokenizerConfig:
    with open(path, "r", encoding="utf-8") as f:
        return TokenizerConfig.from_json_dict(json.load(f))


def save_tokenizer_config(config: TokenizerConfig, path: str) -> None:
    with open(path, "w", encoding="utf-8") as f:
        json.dump(config.to_json_dict(), f, indent=2, sort_keys=False)
        f.write("\n")

File: cortex-standalone/core/tokenizer.ts
/**
 * The tokenizer — the byte-identical contract (docs/CONTRACTS.md §3).
 *
 * Three copies of this algorithm exist on purpose (this TypeScript core, the Python trainer,
 * the browser widget). Tests pin all three against `tests/fixtures/tokenizer.parity.json`.
 *
 *   1. Lowercase (when `lower`).
 *   2. Split into maximal runs of Unicode Letter/Number codepoints; everything else separates.
 *   3. Map each word to `vocab[word]`, else `unkId`.
 *   4. Truncate to `maxLen`, pad with `padId`.
 */

import { readFileSync } from 'node:fs';

export interface TokenizerConfig {
  version: string;
  lower: boolean;
  maxLen: number;
  padId: number;
  unkId: number;
  padToken: string;
  unkToken: string;
  vocab: Record<string, number>;
}

/** The word-character class. Exported by name: a drift test in the widget reads it. */
export const WORD_CHAR_RE = /[\p{L}\p{N}]/u;

/** True for any Unicode Letter or Number codepoint. */
export function isWordChar(ch: string): boolean {
  return WORD_CHAR_RE.test(ch);
}

/** Maximal runs of word characters, iterating by codepoint so surrogate pairs never split. */
export function wordSplit(text: string): string[] {
  const tokens: string[] = [];
  let cur = '';
  for (const ch of text) {
    if (isWordChar(ch)) {
      cur += ch;
    } else if (cur) {
      tokens.push(cur);
      cur = '';
    }
  }
  if (cur) tokens.push(cur);
  return tokens;
}

export interface OffsetToken {
  text: string;
  /** UTF-16 code-unit offsets into the ORIGINAL (un-lowercased) string, `[start, end)`. */
  start: number;
  end: number;
}

/** Same split as `wordSplit`, keeping each token's offsets into the original-cased text (§3.5). */
export function wordSplitWithOffsets(text: string): OffsetToken[] {
  const tokens: OffsetToken[] = [];
  let cur = '';
  let start = -1;
  let i = 0;
  for (const ch of text) {
    if (isWordChar(ch)) {
      if (start === -1) start = i;
      cur += ch;
    } else if (cur) {
      tokens.push({ text: cur, start, end: i });
      cur = '';
      start = -1;
    }
    i += ch.length;
  }
  if (cur) tokens.push({ text: cur, start, end: i });
  return tokens;
}

/** text -> exactly `config.maxLen` ids (truncated / padded), OOV words -> `config.unkId`. */
export function tokenize(text: string, config: TokenizerConfig): number[] {
  const lowered = config.lower ? text.toLowerCase() : text;
  const words = wordSplit(lowered);
  // Own-property lookup only: a word like "constructor" must map to unkId, never to an
  // inherited Object.prototype member (the Python side is `dict.get`, the widget hasOwnProperty).
  const ids = words
    .slice(0, config.maxLen)
    .map((w) => (Object.prototype.hasOwnProperty.call(config.vocab, w) ? config.vocab[w] : config.unkId));
  while (ids.length < config.maxLen) ids.push(config.padId);
  return ids;
}

/** Load a `*.tokenizer.json` (§3.3) from disk, checking the shape. */
export function loadTokenizerConfig(path: string): TokenizerConfig {
  const parsed = JSON.parse(readFileSync(path, 'utf8')) as Partial<TokenizerConfig> | null;
  if (
    !parsed ||
    typeof parsed !== 'object' ||
    typeof parsed.vocab !== 'object' ||
    parsed.vocab === null ||
    typeof parsed.maxLen !== 'number' ||
    typeof parsed.padId !== 'number' ||
    typeof parsed.unkId !== 'number' ||
    typeof parsed.lower !== 'boolean'
  ) {
    throw new Error(`loadTokenizerConfig: malformed tokenizer config at ${path}`);
  }
  return parsed as TokenizerConfig;
}
